lang/interpolate/parse.rl

```diff
@@ -0,0 +1,180 @@
+
+		// Index in data where the currently captured string started.
+		idx int
+
+		x string   // The string we use for holding a temporary value.
+		l Literal  // The string literal being read, if any.
+		v Variable // The variable being read, if any.
+
+		// Current token. This is either the variable that we just read
+		// or the string literal. We will append it to `out` and move
+		// on.
+		t Token
+	)
+
+	%%{
+		# Record the current position as the start of a string. This is
+		# usually used with the entry transition (>) to start capturing
+		# the string when a state machine is entered.
+		#
+		# fpc is the current position in the string (basically the same
+		# as the variable `p` but a special Ragel keyword) so after
+		# executing `start`, data[idx:fpc+1] is the string from when
+		# start was called to the current position (inclusive).
+		action start { idx = fpc }
+
+		# TODO: check this !
+		# A variable always starts with an alphabetical char and
+		# contains alphanumeric characters or numbers, underscores, and
+		# non-consecutive dots.
+		var_name
+			= ( [a-z] ([a-z0-9_]
+			  | ('.' | '_') [a-z0-9_])*
+			  )
+			>start
+			@{ v.Name = data[idx:fpc+1] }
+			;
+
+		# var is a reference to a variable.
+		var = '${' var_name '}'
+			;
+
+		#escape_sequence = ( '\\a'
+		#		  | '\\b'
+		#		  | '\\f'
+		#		  | '\\n'
+		#		  | '\\r'
+		#		  | '\\t'
+		#		  | '\\v'
+		#		  )
+		#	>start
+		#	@{ l = Literal{Value: "XXX" } }
+		#	;
+
+		# Anything followed by a '\' is matched and used appropriately.
+		#escaped_lit = '\\' any @{ l = Literal{Value: data[fpc:fpc+1]} };
+		#//l = Literal{Value: "Z("+data[fpc:fpc+1]+"):["+ data[fpc-1:fpc+1] +"]" }
+
+		# works!
+		#escaped_lit = '\\' any
+		#@{
+		#	switch s := data[fpc:fpc+1]; s {
+		#	case "a":
+		#		x = "\a"
+		#	case "b":
+		#		x = "\b"
+		#	//case "e":
+		#	//	x = "\e" // non-standard
+		#	case "f":
+		#		x = "\f"
+		#	case "n":
+		#		x = "\n"
+		#	case "r":
+		#		x = "\r"
+		#	case "t":
+		#		x = "\t"
+		#	case "v":
+		#		x = "\v"
+		#	case "\\":
+		#		x = "\\"
+		#	case "\"":
+		#		x = "\""
+		#	case "$":
+		#		x = "$"
+		#	//case "0":
+		#	//	x = "\x00"
+		#	default:
+		#		//panic(fmt.Sprintf("unhandled escaped: %s", s))
+		#		x = s // XXX: should we error cleanly instead?
+		#	}
+		#	l = Literal{Value: x}
+		#};
+
+		# Any special escape characters are matched here.
+		escaped_lit = '\\' ( 'a' | 'b' | 'f' | 'n' | 'r' | 't' | 'v' | '\\' | '"' | '$' )
+		@{
+			switch s := data[fpc:fpc+1]; s {
+			case "a":
+				x = "\a"
+			case "b":
+				x = "\b"
+			//case "e":
+			//	x = "\e" // non-standard
+			case "f":
+				x = "\f"
+			case "n":
+				x = "\n"
+			case "r":
+				x = "\r"
+			case "t":
+				x = "\t"
+			case "v":
+				x = "\v"
+			case "\\":
+				x = "\\"
+			case "\"":
+				x = "\""
+			case "$":
+				x = "$"
+			//case "0":
+			//	x = "\x00"
+			default:
+				//panic(fmt.Sprintf("unhandled escaped: %s", s))
+				x = s // XXX: should we error cleanly instead?
```

Comment:
It makes sense to raise an `error` of `invalid escape sequence '$rune'` or similar. That's what most compilers would do IMO

Reply:
Mostly because a user might be expecting a particular behaviour and silent failures are a nightmare to debug